### PR TITLE
Feature/benchmark

### PR DESCRIPTION
**Pull request by maochuanli**
Benchmark support

0, On the specific comment for a PR:  "@benchmark_bot", the jenkins pipleline will be triggered
1, Jenkins job will checkout the source code from the PR branch;
2, Build the code with command:  

> cargo build --release --features runtime-benchmarks

, be noted the extra --features parameter
3, Run 2 tests with commands:
   > cargo test -p crml-staking --features runtime-benchmarks
   > cargo test -p crml-cennzx --features runtime-benchmarks

4, Run the benchmark with command: 
   > ./target/release/cennznet benchmark --chain dev --steps 50 --repeat 20 --pallet "*" --extrinsic "*" --raw --execution=wasm --wasm-execution=compiled --output output_dir

5, Create a clean_dir directory, and check out the source code again into it
6, Copy the above generated files into `runtime/src/weights/` directory
7, Push back the new files

> git add .; git commit -m "commit new benchmark files"; git push

8, Done

### CODE REVIEW

**Comment by maochuanli** [comment]
@benchmark_bot

**Comment by maochuanli** [comment]
@benchmark_bot

**Comment by maochuanli** [comment]
@benchmark_bot

**Comment by maochuanli** [comment]
@benchmark_bot

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+            }
+        }
+
+
+        stage('Run Benchmarks') {
+            agent { label 'benchmark'}
+            steps{
+                sh 'rm -rf output_dir && mkdir output_dir'
+                sh './target/release/cennznet benchmark --chain dev --steps 50 --repeat 20 --pallet "*" --extrinsic "*" --raw --execution=wasm --wasm-execution=compiled --output output_dir'
```
Comment (holygits) [inline Jenkinsfile:45]: @maochuanli recommend change this value to `--repeat 100` or higher to help even the variation in results

**Comment by maochuanli** [comment]
@benchmark_bot

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		sh 'cargo test -p crml-staking --features runtime-benchmarks'
+                sh 'cargo test -p crml-cennzx --features runtime-benchmarks'
+
+            }
+        }
+
+
+        stage('Run Benchmarks') {
+            agent { label 'benchmark'}
+            steps{
```
Comment (jordy25519) [inline Jenkinsfile]:
```suggestion
            steps {
```

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                sh 'rm -rf output_dir && mkdir output_dir'
+                sh './target/release/cennznet benchmark --chain dev --steps 50 --repeat 20 --pallet "*" --extrinsic "*" --raw --execution=wasm --wasm-execution=compiled --output output_dir'
```
Comment (jordy25519) [inline Jenkinsfile]:
```suggestion
                sh './target/release/cennznet benchmark --chain dev --steps 50 --repeat 100 --pallet "*" --extrinsic "*" --raw --execution=wasm --wasm-execution=compiled --output output_dir'
```

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                    image 'rustlang/rust:nightly'
+                    args '-u root:root'
+                }
+            }
+            steps {
+                sh 'mkdir clean_dir && chmod 777 clean_dir'
+                dir('clean_dir'){
+                    checkout([$class: 'GitSCM', branches: [[name: '${CHANGE_BRANCH}']], extensions: [], userRemoteConfigs: [[url: 'git@github.com:cennznet/cennznet.git']]])
+                    sh 'git checkout ${CHANGE_BRANCH}'
+		    sh 'git branch'
```
Comment (jordy25519) [inline Jenkinsfile]: spaces only
```suggestion
                    sh 'git checkout ${CHANGE_BRANCH}'
                    sh 'git branch'
```

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                  rustup uninstall nightly && \
+                  rustup install $RUST_VERSION && \
+                  rustup install $RUST_NIGHTLY && \
+                  rustup default $RUST_VERSION && \
+                  rustup target add --toolchain $RUST_NIGHTLY wasm32-unknown-unknown && \
+                  rustup target add --toolchain $RUST_VERSION x86_64-unknown-linux-musl && \
+                  mv /usr/local/rustup/toolchains/nightly* /usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu
+                '''
+                sh 'cargo build --release --features runtime-benchmarks'
+		sh 'cargo test -p crml-staking --features runtime-benchmarks'
```
Comment (jordy25519) [inline Jenkinsfile]: whitespace
```suggestion
                sh 'cargo build --release --features runtime-benchmarks'
                sh 'cargo test -p crml-staking --features runtime-benchmarks'
```

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,86 @@
+pipeline {
+    agent none
+    environment {
+        RUST_VERSION='1.50.0'
+        RUST_NIGHTLY='nightly-2021-02-21'
+    }
+    options{
+        lock('singleton-build')
```
Comment (jordy25519) [inline Jenkinsfile:8]: so this ensures only one benchmark will run at a time?

**Review by jordy25519 — APPROVED** [review]
lgtm! just some formatting nits

File: Jenkinsfile (hunk 6 above)
Comment (maochuanli) [reply Jenkinsfile:8]: yes

File: Jenkinsfile (hunk 5 above)
Comment (maochuanli) [reply Jenkinsfile]: I will remove those spaces in a new commit.

File: Jenkinsfile (hunk 3 above)
Comment (maochuanli) [reply Jenkinsfile]: ok 100 times

**Comment by jordy25519** [comment]
@benchmark_bot

**Comment by maochuanli** [comment]
run benchmark

**Review by aliXsed — APPROVED** [review]

**Comment by jordy25519** [comment]
Will be too complicated to sign the commits in this PR, have an issue open to add signing for the bot

**Pull request merged**